### ISSUE TITLE
Issues with DtTrigger when rendering multiple tables

### ISSUE THREAD

**Issue by raydawg2000** [l-lin/angular-datatables#1165]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x] Bug report  
[ ] Feature request
[ ] Documentation issue or request
</code></pre>

## What versions you are using?

<pre><code>
- node version: 6.11.4
- angular version: 5.1.2
- angular-cli version: 1.6.3
- jquery version: 3.2.1
- datatables version: 1.10.16
- angular-datatables version: 5.0.0
</code></pre>

## Current behavior
I am trying to use multiple tables in a single view and used this Issue for reference on setup
https://github.com/l-lin/angular-datatables/issues/1088

I have a dtOptions and dtTrigger in an array, an instance for each table.

```
@ViewChildren(DataTableDirective)
dtElements: QueryList<DataTableDirective>;
dtOptions: DataTables.Settings[] = [];
dtTrigger: Subject<any>[] = [];

ngOnInit() {
    this.dtTrigger["new"] = new Subject<any>();
    this.dtOptions["new"] = { ... };
}

<table datatable [dtOptions]="dtOptions['new']" [dtTrigger]="dtTrigger['new']"></table>
```

The table does not render with options.  If I remove dtTrigger from the table config then the table renders as expected with options.  The issue is then I cannot rerender the table because I need to call 

`dtTrigger.next()`

but dtTrigger does not exist if I don't include it in the table definition.  If I drop the array of dtTriggers and just use a single instance of dtTrigger for testing purposes

```
dtTrigger: Subject<any> = new Subject<any>();

<table datatable [dtOptions]="dtOptions['new']" [dtTrigger]="dtTrigger"></table>
```

This also fails to render the table.  So it appears there is something not working properly with using dtTrigger when using the recommended method of rendering multiple tables.

## Expected behavior

Datatables should function normally, not crash when using DtTrigger

## Minimal reproduction of the problem with instructions
This was a template provided from another post and i modified to show the issue.  Removing the dtTrigger from the table definition allows for table rendering.  Any attempt to use dtTrigger results in the table not rendering.

http://plnkr.co/edit/ntYDaywa9xZ81AQR2obP

**Comment by l-lin** [comment]
You need to wrap the `dtTrigger['new'].next()` with a `setTimeout` so that instruction is "part" of the NgZone:
```ts
setTimeout(() => {
  this.dtTrigger['new'].next();    
});
```

See this [plnkr](http://plnkr.co/edit/Xj3v2RXhBjXglo6wne30?p=preview).

**Comment by vsundesha** [comment]
Worked for me ! Thank you !

> You need to wrap the `dtTrigger['new'].next()` with a `setTimeout` so that instruction is "part" of the NgZone:
> 
> ```ts
> setTimeout(() => {
>   this.dtTrigger['new'].next();    
> });
> ```
> 
> See this [plnkr](http://plnkr.co/edit/Xj3v2RXhBjXglo6wne30?p=preview).

**Comment by mritunjaybs1510** [comment]
Hi I am using the Angular datatables. Multiple table on same page but displaying in two tabs. i have gone through the above discussion and followed the suggestions mentioned for adding the setTimeout(() => {
  this.dtTrigger['new'].next();    
}); 
But this solutions is not worked for me.. Here is my code
**My TypeScript Code**
import { Component, OnInit, Input, ViewChild, OnDestroy, AfterViewInit, EventEmitter, Output, ViewChildren } from '@angular/core';
import { QueryList } from '@angular/core';
import { DataTableDirective } from 'angular-datatables';
import { Subject } from 'rxjs/Subject';
import { TabDirective } from 'ngx-bootstrap/tabs';
import 'rxjs/add/operator/map';
@Component({
  selector: 'app-student-list',
  templateUrl: './student-list.component.html',
  styleUrls: ['./student-list.component.css']
})
export class StudentListComponent implements OnDestroy, OnInit, AfterViewInit {
  @Input('studentList') studentList: any;
  @Input('allottedStudentList') allottedStudentList: any;
  @Output('clickCheckBox') clickCheckBox  = new EventEmitter<any>();
  @Output('selectAllCheckBox') selectAllCheckBox  = new EventEmitter<any>();
  @ViewChildren(DataTableDirective) dtElements: QueryList<DataTableDirective>;
  // @ViewChild(DataTableDirective) dtElement: DataTableDirective;
  dtOptions: DataTables.Settings[] = [];
  // dtTrigger: any = new Subject();
  dtTrigger: Subject<any>[] = [];

  // dtTrigger1: any = new Subject();
  studentIdArray: any = [];
  constructor() { }

  ngOnInit() {
    this.dtOptions[0] = {
      paging: false,
      processing: true,
      search: true,
      searching: true,
      columnDefs: [ {
        'targets': 4,
        'orderable': false
        } ]
    };
    this.dtOptions[1] = {
      paging: false,
      processing: true,
      search: true,
      searching: true,
      columnDefs: [ {
        'targets': 4,
        'orderable': false
        } ]
    };
    this.dtTrigger['first'] = new Subject<any>();
    this.dtTrigger['second'] = new Subject<any>();
  }

  ngOnDestroy(): void {
    // Do not forget to unsubscribe the event
    this.dtTrigger['first'].unsubscribe();
    setTimeout(() => {
      this.dtTrigger['second'].next();
    });
    // this.dtTrigger1.unsubscribe();
  }

  ngAfterViewInit(): void {
    this.dtTrigger['first'].next();
    setTimeout(() => {
      this.dtTrigger['second'].next();
    });
    // this.dtTrigger1.next();
  }

  rerender(): void {
  this.dtElements.forEach((dtElement: DataTableDirective) => {
    dtElement.dtInstance.then((dtInstance: DataTables.Api) => {
      dtInstance.destroy();
      this.dtTrigger['first'].next();
      setTimeout(() => {
        this.dtTrigger['second'].next();
      });

    });
  });
  }

  selectCheckBox(studentId: any, usn: any) {
    if ($('#select_' + studentId).is(':checked')) {
      this.studentIdArray.push(studentId + '-' + usn);
    } else {
      this.studentIdArray.splice(this.studentIdArray.indexOf(studentId + '-' + usn), 1);
    }

   this.clickCheckBox.emit(this.studentIdArray);

  }

  selectAll() {
    const that = this;
     if ($('#select_all').is(':checked')) {
      $('.select_student').each(function() {
        $(this).prop('checked', true);
          if ($(this).is(':checked')) {
            that.studentIdArray.push($(this).attr('student') + '-' + $(this).attr('student_usn'));
          } else {
            that.studentIdArray.splice(that.studentIdArray.indexOf($(this).attr('student') + '-' + $(this).attr('student_usn')), 1);
          }
      });
     } else {
      $('.select_student').each(function() {
        $(this).prop('checked', false);
          if ($(this).is(':checked')) {
            that.studentIdArray.push($(this).attr('student'));
          } else {
            that.studentIdArray.splice(that.studentIdArray.indexOf($(this).attr('student')), 1);
          }
      });
     }

     this.selectAllCheckBox.emit(this.studentIdArray);
    }

}


**HTML CODE**
**table #allottedStudentGrid datatable [dtOptions]="dtOptions[1]" [dtTrigger]="dtTrigger['first']" id="allottedStudentGrid" class="allottedStudentGrid table table-bordered"**

**table #allottedStudentGrid datatable [dtOptions]="dtOptions[1]" [dtTrigger]="dtTrigger['second']" id="allottedStudentGrid" class="allottedStudentGrid table table-bordered**

**Comment by mritunjaybs1510** [comment]
![image](https://user-images.githubusercontent.com/39058426/60574222-d2df1000-9d96-11e9-9724-bb4014ed8b7a.png)

**Comment by mritunjaybs1510** [comment]
Any Help will be greatly appreciated

**Comment by shohag000** [comment]
> Hi I am using the Angular datatables. Multiple table on same page but displaying in two tabs. i have gone through the above discussion and followed the suggestions mentioned for adding the setTimeout(() => {
> this.dtTrigger['new'].next();
> });
> But this solutions is not worked for me.. Here is my code
> **My TypeScript Code**
> import { Component, OnInit, Input, ViewChild, OnDestroy, AfterViewInit, EventEmitter, Output, ViewChildren } from '@angular/core';
> import { QueryList } from '@angular/core';
> import { DataTableDirective } from 'angular-datatables';
> import { Subject } from 'rxjs/Subject';
> import { TabDirective } from 'ngx-bootstrap/tabs';
> import 'rxjs/add/operator/map';
> @component({
> selector: 'app-student-list',
> templateUrl: './student-list.component.html',
> styleUrls: ['./student-list.component.css']
> })
> export class StudentListComponent implements OnDestroy, OnInit, AfterViewInit {
> @input('studentList') studentList: any;
> @input('allottedStudentList') allottedStudentList: any;
> @output('clickCheckBox') clickCheckBox = new EventEmitter();
> @output('selectAllCheckBox') selectAllCheckBox = new EventEmitter();
> @ViewChildren(DataTableDirective) dtElements: QueryList;
> // @ViewChild(DataTableDirective) dtElement: DataTableDirective;
> dtOptions: DataTables.Settings[] = [];
> // dtTrigger: any = new Subject();
> dtTrigger: Subject[] = [];
> 
> // dtTrigger1: any = new Subject();
> studentIdArray: any = [];
> constructor() { }
> 
> ngOnInit() {
> this.dtOptions[0] = {
> paging: false,
> processing: true,
> search: true,
> searching: true,
> columnDefs: [ {
> 'targets': 4,
> 'orderable': false
> } ]
> };
> this.dtOptions[1] = {
> paging: false,
> processing: true,
> search: true,
> searching: true,
> columnDefs: [ {
> 'targets': 4,
> 'orderable': false
> } ]
> };
> this.dtTrigger['first'] = new Subject();
> this.dtTrigger['second'] = new Subject();
> }
> 
> ngOnDestroy(): void {
> // Do not forget to unsubscribe the event
> this.dtTrigger['first'].unsubscribe();
> setTimeout(() => {
> this.dtTrigger['second'].next();
> });
> // this.dtTrigger1.unsubscribe();
> }
> 
> ngAfterViewInit(): void {
> this.dtTrigger['first'].next();
> setTimeout(() => {
> this.dtTrigger['second'].next();
> });
> // this.dtTrigger1.next();
> }
> 
> rerender(): void {
> this.dtElements.forEach((dtElement: DataTableDirective) => {
> dtElement.dtInstance.then((dtInstance: DataTables.Api) => {
> dtInstance.destroy();
> this.dtTrigger['first'].next();
> setTimeout(() => {
> this.dtTrigger['second'].next();
> });
> 
> ```
> });
> ```
> 
> });
> }
> 
> selectCheckBox(studentId: any, usn: any) {
> if ($('#select_' + studentId).is(':checked')) {
> this.studentIdArray.push(studentId + '-' + usn);
> } else {
> this.studentIdArray.splice(this.studentIdArray.indexOf(studentId + '-' + usn), 1);
> }
> 
> this.clickCheckBox.emit(this.studentIdArray);
> 
> }
> 
> selectAll() {
> const that = this;
> if ($('#select_all').is(':checked')) {
> $('.select_student').each(function() {
> $(this).prop('checked', true);
> if ($(this).is(':checked')) {
> that.studentIdArray.push($(this).attr('student') + '-' + $(this).attr('student_usn'));
> } else {
> that.studentIdArray.splice(that.studentIdArray.indexOf($(this).attr('student') + '-' + $(this).attr('student_usn')), 1);
> }
> });
> } else {
> $('.select_student').each(function() {
> $(this).prop('checked', false);
> if ($(this).is(':checked')) {
> that.studentIdArray.push($(this).attr('student'));
> } else {
> that.studentIdArray.splice(that.studentIdArray.indexOf($(this).attr('student')), 1);
> }
> });
> }
> 
> ```
>  this.selectAllCheckBox.emit(this.studentIdArray);
> }
> ```
> 
> }
> 
> **HTML CODE**
> **table #allottedStudentGrid datatable [dtOptions]="dtOptions[1]" [dtTrigger]="dtTrigger['first']" id="allottedStudentGrid" class="allottedStudentGrid table table-bordered"**
> 
> **table #allottedStudentGrid datatable [dtOptions]="dtOptions[1]" [dtTrigger]="dtTrigger['second']" id="allottedStudentGrid" class="allottedStudentGrid table table-bordered**

Hello, ive same issues in multiple data table in single page. Is that problem solved?

**Comment by yeshwantnila** [comment]
Is this problem solved

**Comment by joshi-anshul** [comment]
@shanmukhateja 

any solution?

**Comment by shanmukhateja** [comment]
Hi,

The plnkr link provided is proving difficult to work with even after downloading its contents as zip file. Please provide a new reproducible repo to work, preferably `StackBlitz or a GitHub repo`.

**Comment by CodeEpiphany** [comment]
Also running into something like this.

**Comment by epiphanizer** [comment]
I have <table [dataTable]> within an ngFor* loop and am running into this issue. Most of the examples I've researched are setup with @ViewChildren directives that reference multiple known tables. 

We have an arbitrary amount of Table Singleton components that may be deployed by that loop. What should I do in that use case? 

Is there any documentation for using angular-datatables within a loop?

**Comment by shanmukhateja** [comment]
Hi @epiphanizer could you provide a reproducible GitHub repo or similar? It would help a lot.

**Comment by epiphanizer** [comment]
Hi @shanmukhateja! I ended up figuring out my issue. I will say this for anybody else who runs into this.

We were using ngOnChanges to watch for changes and update the table. ngOnChanges is pretty fickle and was responding to way more changes than we needed... I had to make sure we were setting appropriate conditionals.

The other thing I did was set a "isDtInitalized" parameter within the component class. Combined with an *ngIf on the frontend, that was also helpful.

Before destroying, for some reason, i had to also run dtInstance.clear() before dtInstance.destroy() to actually remove the old data... otherwise, sorts on a changed table would sort the initial data.

Now it works!



Other things --- making sure you are running the .next() call within setTimeout()... establishing a valid subscription and then clearing it, etc.. 

All of these added up to it working just great!
